Moved bulk of Python: Functions info into separate entries (#701)
Codecademy/docs: content/python/concepts/functions/terms/arguments/arguments.md

---
Title: 'Arguments'

> Description: 'Supplies data to a funciton when it is called in a program by order of appearance.'

Reviewer: ```suggestion
Description: 'Supplies data to a function when it is called in a program.'
```

Subjects:
  - 'Computer Science'
  - 'Data Science'
Tags:
  - 'Functions'
  - 'Parameters'
  - 'Arguments'
CatalogContent:
  - 'learn-python-3'
  - 'paths/computer-science'
---

> Unless specified otherwise, arguments passed into a called function are assigned to each parameter in the order in which they appear in the function definition. Thus, they are also known as "positional arguments".

Reviewer: ```suggestion
Arguments supply data to a function when it is called in a program. Unless specified otherwise, arguments passed into a called function are assigned to each parameter in the order in which they appear in the function definition. Thus, they are also known as "positional arguments".
```

## Keyword Arguments

Python also supports keyword arguments — prefixing arguments with the names of parameters to assign them directly, regardless of the order.

```py
def write_a_book(name, color, clothing_item):
  print(name + " was wearing a " + color +
        " " + clothing_item + ".")

write_a_book(color="yellow", clothing_item="raincoat", name="Jonas")
```

The output will look like this:

```shell
Jonas was wearing a yellow raincoat.
```

Keyword arguments must be passed after positional arguments.

```py
write_a_book(name="Jonas", "yellow", "raincoat")
```

The call above raises the following exception:

```error
SyntaxError: positional argument follows keyword argument
```

## Varying Arguments

> When defining a function, it may not be necessary to know in advance how many arguments will be needed. In such cases, a special parameter `*args` is passed in. The asterisk, known in this context as the "packing operator", packs the arguments into a [tuple](https://www.codecademy.com/resources/docs/python/tuples) stored in `args`. This tuple can then be iterated through within the function.

Reviewer: ```suggestion
When defining a function, it may not be known in advance how many arguments will be needed. In such cases, a special parameter `*args` is passed in. The asterisk, known in this context as the "packing operator", packs the arguments into a [tuple](https://www.codecademy.com/resources/docs/python/tuples) stored in `args`. This tuple can then be iterated through within the function.
```

```py
def multiply(*args):
  product = 1
  for arg in args:
    product *= arg
  return product

multiply(21, 24)
multiply(10, 5, 3, 6, 17)
```

The output will look like this:

```shell
504
15300
```

## Varying Keyword Arguments

Similarly, functions can be called with an arbitrary number of keyword arguments. In this case, a special parameter `**kwargs` is passed in, where the double asterisk is a packing operator that produces a [dictionary](https://www.codecademy.com/resources/docs/python/dictionaries) rather than a tuple. The parameter name and value of each keyword argument are packed as a key-value pair stored in `kwargs`.

```py
def north_american_capitals(**kwargs):
  for country in kwargs:
    print(country + ": " + kwargs[country])

north_american_capitals(canada="Ottawa", us="Washington D.C.", mexico="Mexico City")
```

The output of the function call will be:

```shell
canada: Ottawa
us: Washington, D.C.
mexico: Mexico City
```

When defining a function, both forms of argument packing can be used. However, `args` must always precede `kwargs`.

## Passing Arguments Dynamically

> When many arguments need to be passed into a function, it can be tedious to type them out individually. Instead, "argument unpacking" can be used to pass positional or keyword arguments dynamically.

Reviewer: ```suggestion
When many arguments need to be passed into a function, it can be tedious to type them out individually. Instead, "argument unpacking" can be used to pass positional or keyword arguments dynamically. Using the `*` or `**` prefixes respectively.
```
